providers/jetstream/jsprovider.go

```diff
@@ -0,0 +1,367 @@
+	entries, err := p.kv.History(context.Background(), fmt.Sprintf("%s.*", prefix))
+	if err != nil {
+		if err == jetstream.ErrKeyNotFound {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	m := make(map[string][]byte)
+	for _, e := range entries {
+		if e.Operation() != jetstream.KeyValuePut {
+			delete(m, e.Key())
+			continue
+		} else {
+			n := e.Key()[len(prefix)+1:]
+			m[n] = e.Value()
+		}
+	}
+	return m, nil
+}
+
+func (p *Provider) Load() ([]*ab.OperatorData, error) {
+	m, err := p.getChildren(OperatorPrefix)
+	if err != nil {
+		return nil, err
+	}
+	operators := make([]*ab.OperatorData, 0, len(m))
+	for _, v := range m {
+		o := &ab.OperatorData{
+			BaseData: ab.BaseData{
+				Token: string(v),
+			},
+		}
+		oc, err := jwt.DecodeOperatorClaims(o.Token)
+		if err != nil {
+			return nil, err
+		}
+		o.Claim = oc
+		o.Loaded = o.Claim.IssuedAt
+		o.EntityName = o.Claim.Name
+		o.Key, err = p.getKey(o.Claim.Subject)
+		if err != nil {
+			return nil, err
+		}
+		for _, sk := range o.Claim.SigningKeys {
+			k, err := p.getKey(sk)
+			if err != nil {
+				return nil, err
+			}
+			o.OperatorSigningKeys = append(o.OperatorSigningKeys, k)
+		}
+		if err := p.loadAccounts(o); err != nil {
+			return nil, err
+		}
+		operators = append(operators, o)
+	}
+	return operators, nil
+}
+
+func (p *Provider) loadAccounts(od *ab.OperatorData) error {
+	// accounts stored under <operatorPublicKey>.<accountPublicKey>
+	m, err := p.getChildren(od.Claim.Subject)
+	if err != nil {
+		return err
+	}
+	for _, v := range m {
+		a := &ab.AccountData{
+			Operator: od,
+			BaseData: ab.BaseData{
+				Token: string(v),
+			},
+		}
+		ac, err := jwt.DecodeAccountClaims(a.Token)
+		if err != nil {
+			return err
+		}
+		a.Claim = ac
+		a.Loaded = a.Claim.IssuedAt
+		a.EntityName = a.Claim.Name
+		a.Key, err = p.getKey(a.Claim.Subject)
+		if err != nil {
+			return err
+		}
+		for pk := range a.Claim.SigningKeys {
+			k, err := p.getKey(pk)
+			if err != nil {
+				return err
+			}
+			a.AccountSigningKeys = append(a.AccountSigningKeys, k)
+		}
+		if err := p.loadUsers(a); err != nil {
+			return err
+		}
+		od.AccountDatas = append(od.AccountDatas, a)
+	}
+	return nil
+}
+
+func (p *Provider) loadUsers(ad *ab.AccountData) error {
+	// users stored under <accountPublicKey>.<userPublicKey>
+	m, err := p.getChildren(ad.Claim.Subject)
+	if err != nil {
+		return err
+	}
+	for _, v := range m {
+		u := &ab.UserData{
+			AccountData: ad,
+			BaseData: ab.BaseData{
+				Token: string(v),
+			},
+		}
+		uc, err := jwt.DecodeUserClaims(u.Token)
+		if err != nil {
+			return err
+		}
+		u.Claim = uc
+		u.Loaded = u.Claim.IssuedAt
+		u.EntityName = u.Claim.Name
+		u.Key, err = p.getKey(u.Claim.Subject)
+		if err != nil {
+			return err
+		}
+
+		ad.UserDatas = append(ad.UserDatas, u)
+	}
+	return nil
+}
+
+func (p *Provider) getKey(pk string) (*ab.Key, error) {
+	e, err := p.kv.Get(context.Background(), fmt.Sprintf("keys.%s", pk))
```

Comment:
Do you think we should encrypt these keys before putting them in kv? I wish we had the codecs we planned for KV sigh so this whole store could just be encrypted

Reply:
I think if they provide a key to the provider that it should do that.